src/ore/src/operators_async_ext.rs

```diff
@@ -166,23 +177,35 @@ impl<G: Scope> OperatorBuilderExt<G> for OperatorBuilder<G> {
                 if had_pending_notify && !scheduler.is_notified() {
                     waker.wake_by_ref();
                 }
+
+                *reschedule_flag.borrow()
             }
         });
     }
 }
 
-#[allow(unused_macros)]
+/// Convenience macro used to wrap what might otherwise be the argument to `build_reschedule`.
+#[macro_export]
 macro_rules! async_op {
     (|$capabilities:ident, $frontiers:ident| $body:block) => {
-        move |mut capabilities, mut frontiers, scheduler| async move {
+        move |mut capabilities, mut frontiers, scheduler, reschedule_flag| async move {
             loop {
                 scheduler.notified().await;
                 // rebind to mutable references to make sure they can't be accidentally dropped
                 #[allow(unused_mut)]
                 let mut $capabilities = &mut capabilities;
                 #[allow(unused_mut)]
                 let mut $frontiers = &mut frontiers;
-                let _: () = async { $body }.await;
+
+                {
+                    // If this async block is `poll`ed and returns while `$body` is executing, it
+                    // should always be rescheduled.
+                    *reschedule_flag.borrow_mut() = true
```

Comment:
I'd like a second set of eyes on this in particular to make sure I'm correct

Reply:
won't a poll that causes the block to execute a return always return that value? do we need to set this beforehand, or just below?

Reply:
My futures recollection is a little rusty so I may not be quite correct but my understanding is that the `poll` at line 166 can return while `$body` below is in the middle of executing.  In that case, we'd want to unconditionally return `true` from the closure argument of `build_reschedule` (by reading `reschedule_flag`).  All this to tell timely that if the operator is interrupted in the middle of processing an iteration of `$body`, it should always be considered incomplete